Partly fixes #2062: part5b refactor register provider

KSchopmeyer opened the pull request:
See commit message for details.

Marked WIP for the moment but it passes tests.

**DISCUSSION:** In this code I ignore  attribute error in call to post_register_setup(conn).  Not sure whether we should ignore (they supply if they want to) or error(they must supply at least a dummy). Since this is server side code, I figured that they will figure it out.

andy-maier commented:
Rebase on master

andy-maier commented:
Please rebase on latest master, deleting the commits that belong to the other PR, and the conflicts will go away.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -307,6 +311,19 @@ class must contain variables `provider_type` and
+                provider.post_register_setup(conn)
+            except AttributeError:
+                if verbose:
+                    _format("Register_Provider post_register_setup not found")
```
I don't think we should require that a provider has this method. It should be optional: If defined, it is called.

KSchopmeyer replied:
Good.  No warning, etc. if missing.

KSchopmeyer commented:
I rebased on master and cleaned up from where I had multiple commits of my own.  Also rewrote the commit message.

Sorry but I cannot figure out how to  remove the two old commits.  Rebasing on master does not do it. Suggestions welcome.

Note that we simply try the post_register_setup and ignore if there is an attribute error.  Since this is a mocker, if there are other errors, it should probably fail or at least capture those exceptions so I do not unregister the provider upon error.

andy-maier commented:
I squashed the 3 commits into one and will merge it when the tests complete.
